crates/turbo-tasks/src/native_function.rs

```diff
@@ -1,18 +1,98 @@
-use std::{fmt::Debug, hash::Hash};
+use std::{fmt::Debug, hash::Hash, pin::Pin};
 
+use anyhow::{Context, Result};
+use futures::Future;
+use serde::{Deserialize, Serialize};
 use tracing::Span;
 
 use crate::{
     self as turbo_tasks,
+    magic_any::{MagicAny, MagicAnyDeserializeSeed, MagicAnySerializeSeed},
     registry::register_function,
     task::{
-        function::{IntoTaskFnWithThis, NativeTaskFn},
+        function::{IntoTaskFnWithThis, NativeTaskFuture},
         IntoTaskFn, TaskFn,
     },
-    util::SharedError,
-    ConcreteTaskInput, RawVc,
+    RawVc, TaskInput,
 };
 
+type ResolveFunctor =
+    for<'a> fn(
+        &'a dyn MagicAny,
+    ) -> Pin<Box<dyn Future<Output = Result<Box<dyn MagicAny>>> + Send + 'a>>;
+
+type IsResolvedFunctor = fn(&dyn MagicAny) -> bool;
+
+pub struct ArgMeta {
+    serializer: MagicAnySerializeSeed,
+    deserializer: MagicAnyDeserializeSeed,
+    is_resolved: IsResolvedFunctor,
+    resolve: ResolveFunctor,
+}
+
+impl ArgMeta {
+    pub fn new<T>() -> Self
+    where
+        T: TaskInput + Serialize + for<'de> Deserialize<'de> + 'static,
+    {
+        Self {
+            serializer: MagicAnySerializeSeed::new::<T>(),
+            deserializer: MagicAnyDeserializeSeed::new::<T>(),
+            is_resolved: |value| {
+                value
+                    .downcast_ref::<T>()
+                    .with_context(|| {
+                        #[cfg(debug_assertions)]
+                        return format!(
+                            "Invalid argument type, expected {} got {}",
+                            std::any::type_name::<T>(),
+                            value.magic_type_name()
+                        );
+                        #[cfg(not(debug_assertions))]
+                        return "Invalid argument type";
+                    })
+                    .unwrap()
```

Comment:
This downcasting bit with pretty error messages is the same between both `resolve` and `is_resolved`? Move it into a generic helper method?

Also, thanks for minding the binary size here with `debug_assertions`!